crates/matrix-sdk-crypto/src/store/memorystore.rs

```diff
@@ -634,3 +634,263 @@ mod tests {
         assert!(store.is_message_known(&hash).await.unwrap());
     }
 }
+
+#[cfg(test)]
+mod integration_tests {
+    use std::{
+        collections::HashMap,
+        sync::{Arc, Mutex, OnceLock},
+    };
+
+    use async_trait::async_trait;
+    use ruma::{
+        events::secret::request::SecretName, DeviceId, OwnedDeviceId, RoomId, TransactionId, UserId,
+    };
+
+    use super::MemoryStore;
+    use crate::{
+        cryptostore_integration_tests, cryptostore_integration_tests_time,
+        olm::{
+            InboundGroupSession, OlmMessageHash, OutboundGroupSession, PrivateCrossSigningIdentity,
+            StaticAccountData,
+        },
+        store::{BackupKeys, Changes, CryptoStore, PendingChanges, RoomKeyCounts, RoomSettings},
+        types::events::room_key_withheld::RoomKeyWithheldEvent,
+        Account, GossipRequest, GossippedSecret, ReadOnlyDevice, ReadOnlyUserIdentities,
+        SecretInfo, Session, TrackedUser,
+    };
+
+    /// Holds on to a MemoryStore during a test, and moves it back into STORES
+    /// when this is dropped
+    #[derive(Clone, Debug)]
+    struct PersistentMemoryStore(Arc<MemoryStore>);
+
+    impl PersistentMemoryStore {
+        fn new() -> Self {
+            Self(Arc::new(MemoryStore::new()))
+        }
+
+        fn get_static_account(&self) -> Option<StaticAccountData> {
+            self.0.get_static_account()
+        }
+    }
+
+    impl MemoryStore {
+        fn get_static_account(&self) -> Option<StaticAccountData> {
+            self.account.read().unwrap().as_ref().map(|acc| acc.static_data().clone())
+        }
+    }
+
+    /// Return a clone of the store for the test with the supplied name. Note: dropping this store
+    /// won't destroy its data, since [PersistentMemoryStore] is a reference-counted smart pointer
+    /// to an underlying [MemoryStore].
+    async fn get_store(name: &str, _passphrase: Option<&str>) -> PersistentMemoryStore {
+        // Holds on to one [PersistentMemoryStore] per test, so even if the test drops the store, we
+        // keep its data alive. This simulates the behaviour of the other stores, which keep their
+        // data in a real DB, allowing us to test MemoryStore using the same code.
+        static STORES: OnceLock<Mutex<HashMap<String, PersistentMemoryStore>>> = OnceLock::new();
+        let stores = STORES.get_or_init(|| Mutex::new(HashMap::new()));
+
+        stores
+            .lock()
+            .unwrap()
+            .entry(name.to_owned())
+            .or_insert_with(|| PersistentMemoryStore::new())
+            .clone()
+    }
+
+    /// Forwards all methods to the underlying [MemoryStore].
+    #[async_trait]
+    impl CryptoStore for PersistentMemoryStore {
```

Comment:
Instead of reimplementing the full trait, any chance we can just have `impl Deref for PersistentMemoryStore`, with a target set to `&self.0`?

Extending that proposal: do we even need the `PersistentMemoryStore` wrapper at all? It seems that it could be an `Arc<MemoryStore>` directly, and `Deref` on `Arc` would automatically do what we want here :eyes: